`crl test` fails to create Python 3.10 based virtualenvs

The following error is raised by `crl test` when trying to create Python 3.10 virtualenv:

> Traceback (most recent call last):
  File "/home/jenkins/workspace/Merge/tox-test/.tox/test/lib/python3.7/site-packages/virtualenv.py", line 2502, in <module>
    main()
  File "/home/jenkins/workspace/Merge/tox-test/.tox/test/lib/python3.7/site-packages/virtualenv.py", line 783, in main
    create_environment(
  File "/home/jenkins/workspace/Merge/tox-test/.tox/test/lib/python3.7/site-packages/virtualenv.py", line 1070, in create_environment
    install_python(home_dir, lib_dir, inc_dir, bin_dir, site_packages=site_packages, clear=clear, symlink=symlink)
  File "/home/jenkins/workspace/Merge/tox-test/.tox/test/lib/python3.7/site-packages/virtualenv.py", line 1281, in install_python
    fix_lib64(lib_dir, symlink)
  File "/home/jenkins/workspace/Merge/tox-test/.tox/test/lib/python3.7/site-packages/virtualenv.py", line 1702, in fix_lib64
    assert os.path.basename(lib_dir) == "python{}".format(sys.version[:3]), "Unexpected python lib dir: {!r}".format(
AssertionError: Unexpected python lib dir: '/home/jenkins/workspace/Merge/tox-test/.tox/py310/lib/python3.10'
Running virtualenv with interpreter /usr/bin/python3.10


Run `tox -e test` in a directory containing the following `tox.ini` to demonstrate the problem:

```
[tox]
envlist = py{37,310}
skipsdist = True

[testenv]
basepython =
    py310: python3.10
install_command = pip install --no-cache-dir --timeout 60 {opts} {packages}
commands = python --version

[testenv:test]
basepython = python3.7
deps =
    tox
    crl.devutils
commands =
    python --version
    tox -e py310
```

Root cause is probably virtualenv==16.3.0 required by version 1.2.4

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -3,6 +3,14 @@
 CHANGES
 =======
 
+1.2.6
+-----
+
+- Support for Python versions 3.8, 3.9 and 3.10
+- Use importlib in setup.py for Python 3.x (instead of deprecated imp module)
+- Require Jinja2==3.0.3 to avoid errors caused by removal of 'contextfunction'
+- Require pytest-flake8==1.0.7 and lazy-object-proxy==1.6.0 for Python 2.7 tests
+
 1.2.5
 -----
 

--- a/setup.py
+++ b/setup.py
@@ -45,6 +45,7 @@ setup(
                       'six',
                       'rstcheck',
                       'sphinx < 1.8.0',
+                      'Jinja2==3.0.3;python_version>"3.0"',
                       'robotframework',
                       'virtualenvrunner',
                       'virtualenv==16.3.0',

--- a/setup.py
+++ b/setup.py
@@ -48,7 +48,7 @@ setup(
                       'Jinja2==3.0.3;python_version>"3.0"',
                       'robotframework',
                       'virtualenvrunner',
-                      'virtualenv==16.3.0',
+                      'virtualenv==20.13.4',
                       'configparser'],
     long_description=read('README.rst'),
     license='BSD 3-Clause',

--- a/setup.py
+++ b/setup.py
@@ -1,5 +1,5 @@
 import os
-import imp
+import sys
 from setuptools import setup, find_packages
 
 __copyright__ = 'Copyright (C) 2019, Nokia'
@@ -9,8 +9,20 @@ VERSIONFILE = os.path.join(
     'src', 'crl', 'devutils', '_version.py')
 
 
+def import_module(name, path):
+    if sys.version_info.major == 2:
+        import imp
+        return imp.load_source(name, path)
+
+    import importlib
+    spec = importlib.util.spec_from_file_location(name, path)
+    module = importlib.util.module_from_spec(spec)
+    spec.loader.exec_module(module)
+    return module
+
+
 def get_version():
-    return imp.load_source('_version', VERSIONFILE).get_version()
+    return import_module('_version', VERSIONFILE).get_version()
 
 
 def read(fname):

--- a/setup.py
+++ b/setup.py
@@ -2,7 +2,7 @@ import os
 import sys
 from setuptools import setup, find_packages
 
-__copyright__ = 'Copyright (C) 2019, Nokia'
+__copyright__ = 'Copyright (C) 2019-2022, Nokia'
 
 VERSIONFILE = os.path.join(
     os.path.dirname(os.path.abspath(__file__)),
@@ -59,6 +59,10 @@ setup(
                  'Programming Language :: Python :: 3',
                  'Programming Language :: Python :: 3.5',
                  'Programming Language :: Python :: 3.6',
+                 'Programming Language :: Python :: 3.7',
+                 'Programming Language :: Python :: 3.8',
+                 'Programming Language :: Python :: 3.9',
+                 'Programming Language :: Python :: 3.10',
                  'Topic :: Software Development'],
     keywords='robotframework testing testautomation acceptancetesting atdd bdd',
     url='https://github.com/nokia/crl-devutils',

--- a/src/crl/devutils/_version.py
+++ b/src/crl/devutils/_version.py
@@ -1,5 +1,5 @@
 __copyright__ = 'Copyright (C) 2019, Nokia'
-VERSION = '1.2.5'
+VERSION = '1.2.6'
 GITHASH = ''
 
 

--- a/tox.ini
+++ b/tox.ini
@@ -1,6 +1,6 @@
 # Copyright (C) 2019, Nokia
 [tox]
-envlist = py27, py36, py37, docs, pylint, pylint3
+envlist = py{27,36,37,38,39,310}, docs, pylint, pylint3
 
 [base]
 deps =
@@ -26,15 +26,21 @@ setenv =
     COVERAGE_FILE = .coverage{envname}
 passenv = COVERAGE_FILE
 changedir = {envtmpdir}
-commands = {posargs:py.test -v --junitxml=junit.xml \
+commands =
+    {posargs:py.test -v --junitxml=junit.xml \
 	   --cov-config {toxinidir}/.coveragerc \
 	   --cov={envsitepackagesdir}/crl/devutils \
            {envsitepackagesdir}/crl/devutils {toxinidir}/tests}
-
+basepython =
+    py27: python2.7
+    py36: python3.6
+    py37: python3.7
+    py38: python3.8
+    py39: python3.9
+    py310: python3.10
 deps =
     py27: {[base2]deps}
-    py36: {[base3]deps}
-    py37: {[base3]deps}
+    py{36,37,38,39,310}: {[base3]deps}
 
 [pytest]
 addopts = --flake8 --cov-report xml

--- a/tox.ini
+++ b/tox.ini
@@ -6,19 +6,20 @@ envlist = py27, py36, py37, docs, pylint, pylint3
 deps =
     mock
     pytest-cov
-    pytest-flake8
     fixtureresources
 
 [base2]
 deps =
     {[base]deps}
     pytest < 4.6
+    pytest-flake8 == 1.0.7
     more-itertools <= 5.0.0
 
 [base3]
 deps =
     {[base]deps}
     pytest
+    pytest-flake8
 
 [testenv]
 setenv =
@@ -46,6 +47,7 @@ filterwarnings =
 basepython = python2.7
 deps =
     pylint < 2.0
+    lazy-object-proxy == 1.6.0
     {[base2]deps}
 commands = pylint {posargs: --reports=n  --rcfile={toxinidir}/.pylintrc \
                            {toxinidir}/src/crl {toxinidir}/tests}

